Mathlib/Algebra/GroupWithZero/Units/Basic.lean

```diff
@@ -0,0 +1,373 @@
+
+theorem mul_inverse_cancel_left (x y : M₀) (h : IsUnit x) : x * (inverse x * y) = y := by
+  rw [← mul_assoc, mul_inverse_cancel x h, one_mul]
+#align ring.mul_inverse_cancel_left Ring.mul_inverse_cancel_left
+
+theorem inverse_mul_cancel_left (x y : M₀) (h : IsUnit x) : inverse x * (x * y) = y := by
+  rw [← mul_assoc, inverse_mul_cancel x h, one_mul]
+#align ring.inverse_mul_cancel_left Ring.inverse_mul_cancel_left
+
+theorem inverse_mul_eq_iff_eq_mul (x y z : M₀) (h : IsUnit x) : inverse x * y = z ↔ y = x * z :=
+  ⟨fun h1 => by rw [← h1, mul_inverse_cancel_left _ _ h],
+  fun h1 => by rw [h1, inverse_mul_cancel_left _ _ h]⟩
+#align ring.inverse_mul_eq_iff_eq_mul Ring.inverse_mul_eq_iff_eq_mul
+
+theorem eq_mul_inverse_iff_mul_eq (x y z : M₀) (h : IsUnit z) : x = y * inverse z ↔ x * z = y :=
+  ⟨fun h1 => by rw [h1, inverse_mul_cancel_right _ _ h],
+  fun h1 => by rw [← h1, mul_inverse_cancel_right _ _ h]⟩
+#align ring.eq_mul_inverse_iff_mul_eq Ring.eq_mul_inverse_iff_mul_eq
+
+variable (M₀)
+
+@[simp]
+theorem inverse_one : inverse (1 : M₀) = 1 :=
+  inverse_unit 1
+#align ring.inverse_one Ring.inverse_one
+
+@[simp]
+theorem inverse_zero : inverse (0 : M₀) = 0 := by
+  nontriviality
+  exact inverse_non_unit _ not_isUnit_zero
+#align ring.inverse_zero Ring.inverse_zero
+
+variable {M₀}
+
+end Ring
+
+theorem IsUnit.ring_inverse {a : M₀} : IsUnit a → IsUnit (Ring.inverse a)
+  | ⟨u, hu⟩ => hu ▸ ⟨u⁻¹, (Ring.inverse_unit u).symm⟩
+#align is_unit.ring_inverse IsUnit.ring_inverse
+
+@[simp]
+theorem isUnit_ring_inverse {a : M₀} : IsUnit (Ring.inverse a) ↔ IsUnit a :=
+  ⟨fun h => by
+    cases subsingleton_or_nontrivial M₀
+    · convert h
+      -- Porting note:
+      -- This is needed due to a regression in `convert` noted in https://github.com/leanprover-community/mathlib4/issues/739
+      exact Subsingleton.elim _ _
+    · contrapose h
+      rw [Ring.inverse_non_unit _ h]
+      exact not_isUnit_zero
+      ,
+    IsUnit.ring_inverse⟩
+#align is_unit_ring_inverse isUnit_ring_inverse
+
+namespace Units
+
+variable [GroupWithZero G₀]
+
+variable {a b : G₀}
+
+/-- Embed a non-zero element of a `GroupWithZero` into the unit group.
+  By combining this function with the operations on units,
+  or the `/ₚ` operation, it is possible to write a division
+  as a partial function with three arguments. -/
+def mk0 (a : G₀) (ha : a ≠ 0) : G₀ˣ :=
+  ⟨a, a⁻¹, mul_inv_cancel ha, inv_mul_cancel ha⟩
+#align units.mk0 Units.mk0
+
+@[simp]
+theorem mk0_one (h := one_ne_zero) : mk0 (1 : G₀) h = 1 := by
+  ext
+  rfl
+#align units.mk0_one Units.mk0_one
+
+@[simp]
+theorem val_mk0 {a : G₀} (h : a ≠ 0) : (mk0 a h : G₀) = a :=
+  rfl
+#align units.coe_mk0 Units.val_mk0
+
+@[simp]
+theorem mk0_val (u : G₀ˣ) (h : (u : G₀) ≠ 0) : mk0 (u : G₀) h = u :=
+  Units.ext rfl
+#align units.mk0_coe Units.mk0_val
+
+-- porting note: removed `simp` tag because `simpNF` says it's redundant
+theorem mul_inv' (u : G₀ˣ) : (u : G₀) * u⁻¹ = 1 := by
```

Comment:
I don't understand why this lemma even exists; it looks to me like it's vestigial from a refactor. Can we just remove it in lean 3?

Reply:
Oh nevermind, there's an invisible coercion here in lean 3, ` (↑ u)⁻¹`. Is this present in lean 4?

Reply:
Yes.